### PR TITLE
Document the dataflow analysis for bounds widening

### PR DESCRIPTION
**Pull request by mgrang**
We use a dataflow algorithm for bounds widening of nt_array_ptrs. Document the
details of the algorithm.

### CODE REVIEW

File: clang/docs/checkedc/Bounds-Widening-for-Null-Terminated-Arrays.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
+# Bounds Widening for Null-terminated Arrays
+
+## Null-terminated Arrays
+A null-terminated array is a sequence of elements in memory that ends with a
+null terminator. Checked C adds the type `nt_array_ptr<T>` to represent
+pointers to these kinds of arrays. These arrays can be divided into two parts:
+a prefix with bounds followed by a null terminator.
+
+An important property of a null-terminated array is that an element of the
+array can be read, provided the preceeding element is not the null terminator.
```
Comment (kkjeer) [inline clang/docs/checkedc/Bounds-Widening-for-Null-Terminated-Arrays.md]: Typo: `preceeding` -> `preceding`

**Comment by mgrang** [comment]
@dtarditi Katherine has approved this PR. Do you want to review this or should I go ahead and merge it?

File: clang/docs/checkedc/Bounds-Widening-for-Null-Terminated-Arrays.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
+# Bounds Widening for Null-terminated Arrays
+
+## Null-terminated Arrays
+A null-terminated array is a sequence of elements in memory that ends with a
+null terminator. Checked C adds the type `nt_array_ptr<T>` to represent
+pointers to these kinds of arrays. These arrays can be divided into two parts:
+a prefix with bounds followed by a null terminator.
+
+An important property of a null-terminated array is that an element of the
```
Comment (dtarditi) [inline clang/docs/checkedc/Bounds-Widening-for-Null-Terminated-Arrays.md:9]: I don't think this is the right description of the properties of a null-terminated array.  I think it is better to describe the array as a prefix with bounds followed by a sequence of elements, ending with an element that is the null terminator.

Comment (mgrang) [reply clang/docs/checkedc/Bounds-Widening-for-Null-Terminated-Arrays.md]: David, I have updated the description of the properties of a null-terminated array.

**Pull request merged**